src/ffi.rs

```diff
@@ -578,6 +578,180 @@ pub unsafe extern "C" fn context_get_result(
         .unwrap()
 }
 
+#[cfg(feature = "expr_validation")]
+pub const ATC_ROUTER_EXPRESSION_VALIDATE_OK: i64 = 0;
+#[cfg(feature = "expr_validation")]
+pub const ATC_ROUTER_EXPRESSION_VALIDATE_FAILED: i64 = 1;
+#[cfg(feature = "expr_validation")]
+pub const ATC_ROUTER_EXPRESSION_VALIDATE_BUF_TOO_SMALL: i64 = 2;
+
+/// Validate the ATC expression with the schema.
+///
+/// # Arguments
+///
+/// - `atc`: the C-style string representing the ATC expression.
+/// - `schema`: a valid pointer to the [`Schema`] object returned by [`schema_new`].
+/// - `fields_buf`: a buffer to store the used fields.
+/// - `fields_len`: a pointer to the length of the fields buffer.
+/// - `fields_total`: a pointer for saving the total number of the fields.
+/// - `operators`: a pointer for saving the used operators with bitflags.
+/// - `errbuf`: a buffer to store the error message.
+/// - `errbuf_len`: a pointer to the length of the error message buffer.
+///
+/// # Returns
+///
+/// Returns an integer value indicating the validation result:
+/// - ATC_ROUTER_EXPRESSION_VALIDATE_OK(0) if validation is passed.
+/// - ATC_ROUTER_EXPRESSION_VALIDATE_FAILED(1) if validation is failed.
+/// - ATC_ROUTER_EXPRESSION_VALIDATE_BUF_TOO_SMALL(2) if the provided fields buffer is not enough.
+///
+/// If `fields_buf` is null and `fields_len` or `fields_total` is non-null, it will write
```

Comment:
I think we should just be optimistic here and always ask the user to pass in a buffer, 99% of the time reallocate will not be necessary, and we can avoid having to parse the expression twice which could be expensive (especially with Regexes).

Reply:
I've removed the ability to acquire the buffer length. Callers must pass a valid buffer if the calling is for getting the fields in use.
But I keep the `fields_buf` null-able because this interface is `expression_validate` and callers may not always want fields. This will allow us to do more refactoring on the Kong side with this interface. Like getting rid of validating expressions by inserting and then deleting them.